warehouse/query-core/src/main/java/datawave/query/jexl/visitors/whindex/WhindexVisitor.java

```diff
@@ -0,0 +1,953 @@
+        public Multimap<String,JexlNode> usedAndedNodes = LinkedHashMultimap.create();
+    }
+    
+    private WhindexVisitor(ShardQueryConfiguration config, Date beginDate, MetadataHelper metadataHelper) {
+        this(config.getWhindexMappingFields(), config.getWhindexFieldMappings(), beginDate, metadataHelper);
+    }
+    
+    private WhindexVisitor(Set<String> mappingFields, Map<String,Map<String,String>> fieldMappings, Date beginDate, MetadataHelper metadataHelper) {
+        this.mappingFields = mappingFields;
+        this.metadataHelper = metadataHelper;
+        this.valueSpecificFieldMappings = pruneFieldMappings(fieldMappings, beginDate);
+    }
+    
+    /**
+     * Expand all nodes which have multiple dataTypes for the field.
+     *
+     * @param config
+     *            Configuration parameters relevant to our query
+     * @param script
+     *            The jexl node representing the query
+     * @return An expanded version of the passed-in script containing whindex nodes
+     */
+    public static <T extends JexlNode> T apply(T script, ShardQueryConfiguration config, Date beginDate, MetadataHelper metadataHelper) {
+        WhindexVisitor visitor = new WhindexVisitor(config, beginDate, metadataHelper);
+        
+        return apply(script, visitor);
+    }
+    
+    /**
+     * Expand all nodes which have multiple dataTypes for the field.
+     *
+     * @param mappingFields
+     *            The value-specific fields required to create a Whindex
+     * @param fieldMappings
+     *            The value-specific field mappings which can be used to create a Whindex
+     * @param script
+     *            The jexl node representing the query
+     * @return An expanded version of the passed-in script containing whindex nodes
+     */
+    public static <T extends JexlNode> T apply(T script, Set<String> mappingFields, Map<String,Map<String,String>> fieldMappings, Date beginDate,
+                    MetadataHelper metadataHelper) {
+        WhindexVisitor visitor = new WhindexVisitor(mappingFields, fieldMappings, beginDate, metadataHelper);
+        
+        return apply(script, visitor);
+    }
+    
+    @SuppressWarnings("unchecked")
+    private static <T extends JexlNode> T apply(T script, WhindexVisitor visitor) {
+        // if there are any geowave functions with multiple fields, split them up into individual functions
+        script = SplitGeoWaveFunctionVisitor.apply(script, visitor.metadataHelper);
+        
+        // need to flatten the tree so i get all and nodes at the same level
+        script = TreeFlatteningRebuildingVisitor.flatten(script);
+        
+        return (T) script.jjtAccept(visitor, new ExpandData());
+    }
+    
+    private Map<String,Map<String,String>> pruneFieldMappings(Map<String,Map<String,String>> valueSpecificFieldMappings, Date beginDate) {
+        Map<String,Map<String,String>> prunedFieldMappings = new HashMap<>();
+        if (beginDate != null) {
```

Comment:
if this is null can we throw an IllegalArugmentException or IllegalStateException?

Reply:
We only need support for beginDate because in **_our_** current use case the original field predates thje whindex field.  If the whindex field and original field had been created at the same time, beginDate wouldn't be needed at all.  I definitely want to support that case, so I am in favor of keeping this as-is.